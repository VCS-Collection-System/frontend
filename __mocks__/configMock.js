export const getConfig = (URL, configUrl, config) => {
  if(URL.indexOf(configUrl) > -1 ) {
    const country = URL.substring(configUrl.length);
    const countryConfig = config[country] || null;
    return Promise.resolve({data: countryConfig});
  } else {
    return Promise.reject({data: null});
  }
}