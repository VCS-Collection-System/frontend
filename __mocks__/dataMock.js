export const employeeInfo = {
  "employeeId":"mock1",
  "firstName":"Mock",
  "lastName":"Tester",
  "email":"test@mock.com",
  "agencyCode":"Department of Mock Safety",
  "agencyName":"Department of Mock Safety",
  "divisionCode":"App Mocking",
  "divisionName":null,
  "supervisor":null,
  "dateOfBirth":"2021-10-07",
  "alternateEmail":"mock@email.com",
  "age":0,
  "ageInMonth":0,
  "workforceId":"mock2",
  "hr":false
}

export const attachmentObject = {
  "originalFileName":"blob",
  "size":100,
  "contentType":"image/png",
  "s3BucketName":"mockBucket",
  "s3UUID":"mockUUID"
}

export const vaxHistory = {
  "id":0,
  "employee": employeeInfo,
  "submissionDate":"2021-10-04T15:12:14.875",
  "review":{
     "reviewerEmployeeId":"z2",
     "reviewerNotes":"",
     "reviewDate":"2021-10-05T14:04:29.296",
     "reviewOutcome":"ACCEPTED",
     "rejectReason":""
  },
  "submittedBy":null,
  "autoApproved":false,
  "lastUpdatedTime":null,
  "lastUpdatedBy":null,
  "vaccineBrand":"MODERNA",
  "vaccineAdministrationDate":"2021-10-04",
  "fullVaccinatedFlag":false,
  "vaccineShotNumber":2,
  "documentType":"VaccineDocument",
  "attachment": {}
}

const taskEmployeeObject = {
  "employeeId":"mock1",
  "firstName":"Mock",
  "lastName":"Tester",
  "email":"test@mock.com",
  "agencyCode":"Department of Mock Safety",
  "agencyName":"Department of Mock Safety",
  "divisionCode":"App Mocking",
  "divisionName":null,
  "supervisor":null,
  "dateOfBirth": {
    "year": "2021",
    "monthValue": "12",
    "dayOfMonth": "16"
  },
  "alternateEmail":"mock@email.com",
  "age":0,
  "ageInMonth":0,
  "workforceId":"mock2",
  "hr":false
}

export const mockTask = {
  "task-input-data": {
    "documentList": {
      "com.redhat.vcs.model.VaccineDocumentList": {
        "documents": [ {
          "com.redhat.vcs.model.VaccineCardDocument": {
            "employee": {
              "com.redhat.vcs.model.Employee": taskEmployeeObject
            },
            "vaccineBrand": {
              "com.redhat.vcs.model.VaccineBrand": "MODERNA"
            },
            "attachment": {
              "com.redhat.vcs.model.Attachment": attachmentObject
            },
            "vaccineAdministrationDate": {
              "year": "2021",
              "monthValue": "12",
              "dayOfMonth": "16"
            },
            "lotNumber": "mockLot"
          } }
        }
      }
    }
  }
}
