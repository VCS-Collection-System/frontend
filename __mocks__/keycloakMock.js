module.exports = {
    token: "bogustoken", 
    authenticated: false,
    loadUserProfile: jest.fn( ()=> {return Promise.resolve({
        attributes: {
            workforceid: ["mockId"]
        }
    })} ),
    login: jest.fn(),
    logout: jest.fn(),
    hasRealmRole: jest.fn(),
    profile: {
        firstName: "mockFirst",
        lastName: "mockLast",
        email: "mock@email.com",
        attributes: {
            workforceid: ["mockId"],
            agency: ["USA"]
        }
    }
};
