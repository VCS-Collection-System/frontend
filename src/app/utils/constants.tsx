declare global {  
  interface Window {
    REACT_APP_PAM_URL_BASE: string;
    REACT_APP_PAM_CONTAINER_ID: string;
    REACT_APP_KEYCLOAK_COUNTRY: string;
    REACT_APP_KEYCLOAK_EMAIL: string;
    REACT_APP_KEYCLOAK_FIRST_NAME: string;
    REACT_APP_KEYCLOAK_ID: string;
    REACT_APP_KEYCLOAK_LAST_NAME: string;
    REACT_APP_MAINTENANCE_BANNER: boolean;
    REACT_APP_MAINTENANCE_BANNER_TEXT: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    REACT_APP_FEATURE_FLAGS: Array<Record<string, any>>;
    REACT_APP_MAX_DOSES: Record<string, number>;
    REACT_APP_CONSENT_LABEL: string;
    REACT_APP_ALTERNATE_CONSENT_LABEL: string;
    REACT_APP_BLOCKED_COUNTRIES: Array<string>;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    REACT_APP_GLOBAL_CONFIG: Record<string, any>;
  }
}

// PAM urls
export const PAM_BASE_URL = window.REACT_APP_PAM_URL_BASE; 
export const SUBMIT_VAX_URL = PAM_BASE_URL + '/attestation/vax';
export const SUBMIT_TEST_URL = PAM_BASE_URL + '/attestation/covid-test-result';
export const PAM_TASK_URL = PAM_BASE_URL + '/tasks/';
export const RESERVED_TASK_URL = PAM_TASK_URL + 'owners?status=Reserved&user=';
export const INPROGRESS_TASK_URL = PAM_TASK_URL + 'owners?status=InProgress&user=';
export const ALL_TASK_URL = PAM_TASK_URL + 'owners?status=Ready&pageSize=200';
export const PRIORITY_URL = PAM_BASE_URL + '/query/priority-results';
export const PRIORITY_CONFIRM_URL = PAM_BASE_URL + '/query/priority-result/';
export const ATTACHMENT_URL = PAM_BASE_URL + '/attestation/attachment';
export const EMPLOYEE_URL = PAM_BASE_URL + '/query/employee/';
export const EMPLOYEE_NAME_URL = PAM_BASE_URL + '/query/employee-name/';
export const EMPLOYEE_VAX_HISTORY_URL = PAM_BASE_URL + '/query/vax-document/';
export const EMPLOYEE_TEST_HISTORY_URL = PAM_BASE_URL + '/query/test-result-document/';
export const VAX_TASK_DOCUMENT_URL = PAM_BASE_URL + '/query/task/document/';
export const DELETE_VAX_URL = PAM_BASE_URL + '/attestation/delete-vax-record/';
export const CONFIG_URL = PAM_BASE_URL + '/query/country-configs/';
export const SAVE_CONFIG_URL = PAM_BASE_URL + '/attestation/country-configs';

// List of countries allow to use this app
export const BLOCKED_COUNTRIES = window.REACT_APP_BLOCKED_COUNTRIES || []

export const KONAMI_CODE = [ "ArrowUp", "ArrowUp", "ArrowDown", "ArrowDown", "ArrowLeft", "ArrowRight", "ArrowLeft", "ArrowRight", "KeyB", "KeyA"];

// Maintenance Banner
export const MAINTENANCE_BANNER = window.REACT_APP_MAINTENANCE_BANNER || false
export const MAINTENANCE_BANNER_TEXT = window.REACT_APP_MAINTENANCE_BANNER_TEXT || ""

// Keycloak claims
export const KEYCLOAK_COUNTRY = window.REACT_APP_KEYCLOAK_COUNTRY
export const KEYCLOAK_EMAIL = window.REACT_APP_KEYCLOAK_EMAIL
export const KEYCLOAK_FIRST_NAME = window.REACT_APP_KEYCLOAK_FIRST_NAME
export const KEYCLOAK_ID = window.REACT_APP_KEYCLOAK_ID
export const KEYCLOAK_LAST_NAME = window.REACT_APP_KEYCLOAK_LAST_NAME
export const FEATURE_FLAGS = window.REACT_APP_FEATURE_FLAGS || []

// Maximum doses allowed by vaccine manufacturer
export const MAX_DOSES = window.REACT_APP_MAX_DOSES || {}

export const CONSENT_LABEL = window.REACT_APP_CONSENT_LABEL
export const ALTERNATE_CONSENT_LABEL = window.REACT_APP_ALTERNATE_CONSENT_LABEL

// Global country config
export const GLOBAL_CONFIG = window.REACT_APP_GLOBAL_CONFIG || {};

