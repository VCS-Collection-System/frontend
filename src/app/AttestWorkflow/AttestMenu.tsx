import React, {useEffect, useState} from 'react';
import {
  Button,
  Card,
  CardBody,
  CardFooter,
  Stack,
  Text,
} from "@patternfly/react-core";
import { CheckCircleIcon, CogIcon, FileUploadIcon } from '@patternfly/react-icons';
import { useHistory } from 'react-router-dom';
import CardWithTitle from '@app/Components/CardWithTitle';
import Feature, { showFeature } from '@app/Components/Feature';
import { useKeycloak } from "@react-keycloak/web";
import { getEmployeeInfo } from '@app/utils/getEmployeeInfo';

const AttestMenu: React.FunctionComponent = () => {
    const history = useHistory();
    const { keycloak } = useKeycloak();
    const employee = getEmployeeInfo(keycloak);
    const [ isRecovery, setIsRecovery ] = useState(false); 

    useEffect(() => {
      showFeature("recovery", keycloak).then(config => {
        setIsRecovery(config);
      });
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);
    
    return (
      <Card isRounded>
        <CardWithTitle title="Menu" info="Please select from the following:" />
        <CardBody>
          <Stack hasGutter>
            <Button
              icon={<FileUploadIcon />}
              id="submit-vax-card-menu-button"
              isSmall
              onClick={() => history.push("/attest/vax")}
              variant="tertiary"
            >
              <Text>Submit Vaccination Record</Text>
              <Feature name="proofOptional" renderOff={() => (
                <Text>Supporting Document Upload Required</Text>
              )}>
                &nbsp;
              </Feature>
            </Button>
            <Feature name="covidTest" renderOff={() =>
              <Feature name="recovery"> 
                <Button
                  icon={<FileUploadIcon />}
                  id="submit-vax-test-menu-button"
                  isSmall
                  onClick={() => history.push("/attest/test")}
                  variant="tertiary"
                >
                  <Text>Submit Recovery Certificate</Text>
                  <Text>Supporting Document Upload Required</Text>
                </Button>
              </Feature>
            }> 
              <Button
                icon={<FileUploadIcon />}
                id="submit-vax-test-menu-button"
                isSmall
                onClick={() => history.push("/attest/test")}
                variant="tertiary"
              >
                <Text>Submit Test Results{isRecovery ? " or Recovery Certificate" : ""}</Text>
                <Text>Supporting Document Upload Required</Text>
              </Button>
            </Feature>
            {employee.isApprover ? (
              <Feature name="hrDashboard"> 
                <Button
                  icon={<CheckCircleIcon />}
                  id="validate-submissions-button"
                  isSmall
                  onClick={() => history.push("/hrdashboard")}
                  variant="tertiary"
                >
                  <Text>HR Dashboard</Text>
                  <Text>Review or Update Red Hat VCS Records</Text>
                </Button>
              </Feature>
            ) : ""}
            {employee.isSuperApprover ? (
              <Button
                  icon={<CogIcon />}
                  id="settings-button"
                  isSmall
                  onClick={() => history.push("/settings")}
                  variant="tertiary"
                >
                  <Text>Application Settings</Text>
                  <Text>Update Red Hat VCS Country Settings</Text>
                </Button>
            ) : ""}
          </Stack>
        </CardBody>
        <CardFooter>
          <Button 
            variant="plain" 
            id="attest-menu-back-button"
            onClick={() => history.push("/profile")}
          >
            Back
          </Button>
        </CardFooter>
      </Card>
    );
};
export { AttestMenu }
