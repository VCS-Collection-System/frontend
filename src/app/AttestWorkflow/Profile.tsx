import React from "react";
import {
  Alert,
  Button,
  Card,
  CardBody,
  Checkbox,
  Form,
  FormGroup,
  TextInput,
} from "@patternfly/react-core";
import { CheckCircleIcon } from "@patternfly/react-icons";
import CardWithTitle from "@app/Components/CardWithTitle";
import { useForm } from "react-hook-form";
import { useHistory } from "react-router-dom";
import { useKeycloak } from "@react-keycloak/web";
import { getEmployeeInfo } from "@app/utils/getEmployeeInfo";
import { BLOCKED_COUNTRIES } from "@app/utils/constants";

const Profile: React.FunctionComponent = () => {
  const history = useHistory();
  const { handleSubmit } = useForm();
  const { keycloak } = useKeycloak();
  const employee = getEmployeeInfo(keycloak);
  const employeeFirstName = employee.firstName;
  const employeeLastName = employee.lastName;
  const employeeEmail = employee.email;
  const employeeCountry = employee.country;

  const onSubmit = () => {
    history.push("/attestmenu");
  };

  return (
    <Card isRounded>
      {typeof(employeeCountry) === "undefined" || BLOCKED_COUNTRIES.includes(employeeCountry) ? (
        <Alert variant="danger" title="This application is not currently available in your region." style={{marginBottom: "var(--pf-global--gutter--md)"}}>
          Please visit our <a href="/contact">Help</a> page for more information.
        </Alert>
      ) : ""}
      <CardWithTitle
        title="Profile"
        info="The information below is retrieved from your Red Hat user account."
      />
      <CardBody>
        <Form onSubmit={handleSubmit(onSubmit)}>
          <FormGroup
            id="employee-full-name"
            label="Full Name"
            fieldId="profile-form-employee-full-name"
          >
            <TextInput
              aria-label="employee-full-name"
              isDisabled
              type="text"
              value={(employeeFirstName || "") + " " + (employeeLastName || "")}
            />
          </FormGroup>
          <FormGroup
            id="employee-email-display"
            label="Email"
            fieldId="profile-form-email"
          >
            <TextInput
              aria-label="employee-email-display"
              isDisabled
              type="email"
              value={employeeEmail || ""}
            />
          </FormGroup>
          {employee.isApprover && 
            (
              <Checkbox
                id="reviewer-checkbox"
                label="Red Hat VCS Reviewer"
                aria-label="disabled checked checkbox"
                defaultChecked
                isDisabled
              />
          )}
          <FormGroup fieldId="accept-test-buttons">
            <Button 
              icon={<CheckCircleIcon />} 
              variant="primary" 
              type="submit"
              isDisabled={typeof(employeeCountry) === "undefined" || BLOCKED_COUNTRIES.includes(employeeCountry)}
              >
              Accept
            </Button>
          </FormGroup>
        </Form>
      </CardBody>
    </Card>
  );
};

export { Profile };
