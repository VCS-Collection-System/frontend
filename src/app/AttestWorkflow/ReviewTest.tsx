import React, { useState } from 'react';
import {
  Button, 
  Card,
  CardFooter,
  CardBody,
  Form,
  FormGroup,
  TextInput
} from '@patternfly/react-core';
import { ExternalLinkAltIcon, OutlinedPaperPlaneIcon } from '@patternfly/react-icons';
import { useHistory } from 'react-router-dom';
import CardWithTitle from '@app/Components/CardWithTitle';
import { convertBase64ToBlob, dateConvert } from "@app/utils/utils";
import authaxios from "@app/utils/axiosInterceptor";
import { ResponseErrorModal } from '@app/Components/FeedbackModals';
import Feature from "@app/Components/Feature";
import { useKeycloak } from "@react-keycloak/web";
import { getEmployeeInfo } from '@app/utils/getEmployeeInfo';
import { SUBMIT_TEST_URL } from '@app/utils/constants';

const ReviewTest: React.FunctionComponent = () => {
  const [isErrorModalOpen, setIsErrorModalOpen] = useState(false); 
  const [errorState, setErrorState] = useState('');
  const [isFormSubmitting, setIsFormSubmitting] = useState(false);
  const history = useHistory();
  const { keycloak } = useKeycloak();
  const employee = getEmployeeInfo(keycloak);
  const employeeFirstName = employee.firstName;
  const employeeLastName = employee.lastName;
  const employeeID = employee.id;
  const employeeEmail = employee.email;
  const employeeCountry = employee.country;
  const employeeName = (employee.firstName || "") + " " + (employee.lastName || ""); 

  const testImage = sessionStorage.getItem("test_image_file") || "";
  const testImageFilename = sessionStorage.getItem("test_image_filename") || "";
  const altFirstName = sessionStorage.getItem("alt_first_name") || "";
  const altLastName = sessionStorage.getItem("alt_last_name") || "";
  const employeeDOB = sessionStorage.getItem("date_of_birth") || "";
  const covidTestResult = sessionStorage.getItem("test_results") || "";
  const covidTestDate = sessionStorage.getItem("test_date") || "";
  const altName = (altFirstName ? altFirstName : employeeFirstName) + " " +  (altLastName ? altLastName : employeeLastName);

  const handleSubmit = ()=> {
    setIsFormSubmitting(true);
    const testCard = convertBase64ToBlob(testImage)
    const formData = new FormData();

    const employeeInfo = {
      id: employeeID,
      firstName: employeeFirstName,
      lastName: employeeLastName,
      altFirstName: altFirstName,
      altLastName: altLastName,
      dateOfBirth: employeeDOB,
      email: employeeEmail,
      agencyCode: employeeCountry,
      agencyName: employeeCountry,
      divisionCode: ""
    };

    const testInfo = {
      covidTestDate: covidTestDate,
      covidTestResult: covidTestResult.toUpperCase()
    }

    const attachmentExtension = testImageFilename.split('.').pop()?.toLowerCase();
    let attachmentType;

    if(attachmentExtension === "pdf") {
      attachmentType = "application/pdf";
    }
    else if(attachmentExtension === "jpg") {
      attachmentType = "image/jpeg";
    }
    else {
      attachmentType = "image/" + attachmentExtension;
    }

    formData.append('employee', new Blob([JSON.stringify(employeeInfo)], {type: "application/json"}) );
    formData.append('documents', new Blob([JSON.stringify(testInfo)], {type: "application/json"}));
    formData.append("attachment", new Blob([testCard], { type: attachmentType }), "blob."+attachmentExtension);

    authaxios.post(SUBMIT_TEST_URL, formData, {
      headers: {
        'Content-Type': 'multipart/form-data;'
      }
    })
    .then(() => {
      setIsFormSubmitting(false);
      history.push('/thankyou');
    })
    .catch(res => {
      if(res.response) {
        setIsFormSubmitting(false);
        setErrorState(res.response.status);
        setIsErrorModalOpen(true);
      }
    })
  }

  return (
    <Card isRounded>
      <ResponseErrorModal 
        isModalOpen={isErrorModalOpen} 
        setIsModalOpen={setIsErrorModalOpen} 
        errorState={errorState}
      />
      <CardWithTitle
        title="Review Submission"
        info="Review your test result submission and validate that test result image matches displayed data. Please press Submit below to proceed."
      />
      <CardBody>
        <Form>
          <FormGroup label="Person ID" fieldId="review-test-form-employee-id">
            <TextInput
              aria-label="review-test-form-employee-id"
              isDisabled
              type="text"
              value={employeeID}
            />
          </FormGroup>
          <FormGroup
            fieldId="review-test-form-employee-full-name"
            label="Full Name"
          >
            <TextInput
              aria-label="review-test-form-employee-full-name"
              isDisabled
              type="text"
              value={employeeName}
            />
          </FormGroup>
          <FormGroup
            fieldId="review-test-form-employee-alt-name"
            label="Full Preferred Name (as displayed on test result)"
          >
            <TextInput
              aria-label="review-test-form-employee-alt-name"
              isDisabled
              type="text"
              value={altName}
            />
          </FormGroup>
          <Feature name="dateOfBirth">
            <FormGroup
              label="Date of Birth"
              fieldId="review-test-date-of-birth"
            >
              <TextInput
                aria-label="review-test-date-of-birth"
                isDisabled
                type="text"
                value={dateConvert(employeeDOB)}
              />
            </FormGroup>
          </Feature>
          <FormGroup label="Test Results" fieldId="review-test-form-results">
            <TextInput
              aria-label="review-test-form-results"
              isDisabled
              type="text"
              value={sessionStorage.getItem("test_results") || ""}
            />
          </FormGroup>
          <FormGroup label="Test Date" fieldId="review-test-form-date">
            <TextInput
              aria-label="review-test-form-date"
              isDisabled
              type="text"
              value={dateConvert(sessionStorage.getItem("test_date") || "")}
            />
          </FormGroup>
          <FormGroup label="Image preview" fieldId="review-vax-image-preview">
            {testImage && testImageFilename.split('.').pop()?.toLowerCase() == "pdf" &&
              (typeof testImage !== "undefined" ? (
                <Button
                  icon={<ExternalLinkAltIcon />}
                  variant="link"
                  component="a"
                  href={URL.createObjectURL(convertBase64ToBlob(testImage))}
                  target="_blank"
                >
                  Test Result PDF
                </Button>
              ) : null)}
            {testImage && testImageFilename.split('.').pop()?.toLowerCase() != "pdf" &&
              (typeof testImage !== "undefined" ? (
                <div>
                  <img
                    src={URL.createObjectURL(convertBase64ToBlob(testImage))}
                    alt="test-results-image"
                  />
                </div>
              ) : null)}
          </FormGroup>
        </Form>
      </CardBody>
      <CardFooter>
        <Button
          icon={<OutlinedPaperPlaneIcon />}
          isDisabled={isFormSubmitting}
          isLoading={isFormSubmitting}
          onClick={handleSubmit}
          variant="primary"
        >
          Submit
        </Button>
        <Button onClick={() => history.push("/attest/test")} variant="plain">
          Back
        </Button>
      </CardFooter>
    </Card>
  );
};

export { ReviewTest };
