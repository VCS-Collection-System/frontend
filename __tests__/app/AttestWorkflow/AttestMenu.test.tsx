import React from "react"
import { shallow, mount } from "enzyme"
import { AttestMenu } from "../../../src/app/AttestWorkflow/AttestMenu"
import mockKeycloak from "../../../__mocks__/keycloakMock"
import { getConfig } from "../../../__mocks__/configMock"
import { CONFIG_URL } from "@app/utils/constants"

jest.mock("@react-keycloak/web", () => ({
  useKeycloak: jest.fn(() => ({ keycloak: mockKeycloak })),
}))

jest.mock("@app/utils/constants", () => ({
  KEYCLOAK_FIRST_NAME: "firstName",
  KEYCLOAK_LAST_NAME: "lastName",
  KEYCLOAK_EMAIL: "email",
  KEYCLOAK_COUNTRY: "attributes.agency[0]",
  KEYCLOAK_ID: "attributes.workforceid[0]",
  APPROVER_ROLES: ["mock-approver"],
  FEATURE_FLAGS: [  
    { 
      "name": "hrdashboard",
      "active": "true"
    }, 
    {
      "name": "covidTest",
      "active": "true"
    }
  ],
  CONFIG_URL: "config/"
}))

const config = {"default": {
  "allowPdf": false,
  "dateOfBirth": true,
  "consentCheckbox": false,
  "covidTest": true,
  "customInput": true,
  "hrDashboard": true,
  "hrSearch": true,
  "cdc": false, 
  "divoc": false, 
  "greenPass": false,
  "recovery": false,
  "fakeFeature": true
}};

jest.mock('@app/utils/axiosInterceptor', () => ({
  get: (URL: string) => getConfig(URL, CONFIG_URL, config)
}))

// mock history.push
import reactRouterDom from "react-router-dom"
const mockPush = jest.fn()
reactRouterDom.useHistory = jest.fn().mockReturnValue({ push: mockPush })
jest.mock("react-router-dom")

describe('AttestWorkflow - AttestMenu', () => {
  it('should be defined', () => {
    expect(AttestMenu).toBeDefined()
  })
  it('should render without crashing', () => {
    const wrapper = shallow(<AttestMenu />)
    expect(wrapper)
  })
  it('should match existing snapshot', () => {
    const wrapper = shallow(<AttestMenu />)
    expect(wrapper).toMatchSnapshot()
  })

  it('Button 0 push to /attest/vax page', async () => {
    const attestMenu = shallow(<AttestMenu />)
    const button = attestMenu.find('#submit-vax-card-menu-button')
    button.simulate('click')
    expect(mockPush).toHaveBeenCalledTimes(1)
    expect(mockPush).toHaveBeenCalledWith('/attest/vax')
  })

  it('Button 1 push to /attest/test page', async () => {
    const attestMenu = mount(<AttestMenu />)
    
    // await new Promise(process.nextTick) doesn't work with 'mount', so use basic timeout instead
    setTimeout(() => {
      const button = attestMenu.find('#submit-vax-test-menu-button').at(0)
      button.simulate('click')
      expect(mockPush).toHaveBeenCalledTimes(1)
      expect(mockPush).toHaveBeenCalledWith('/attest/test')
    }, 500);
    
  })

  it('Button 2 pushes to attest /hrdashboard page', async () => {
    mockKeycloak.hasRealmRole.mockReturnValue(true)
    const attestMenu = mount(<AttestMenu />)
    
    // await new Promise(process.nextTick) doesn't work with 'mount', so use basic timeout instead
    setTimeout(() => {
      const button = attestMenu.find('#validate-submissions-button').at(0)
      button.simulate('click')
      expect(mockPush).toHaveBeenCalledTimes(1)
      expect(mockPush).toHaveBeenCalledWith('/hrdashboard')
    }, 500);
  })

  it('Button 3 pushes to attest /settings page', async () => {
    mockKeycloak.hasRealmRole.mockReturnValue(true)
    const attestMenu = mount(<AttestMenu />)
    const button = attestMenu.find('#settings-button').at(0)
    button.simulate('click')
    expect(mockPush).toHaveBeenCalledTimes(1)
    expect(mockPush).toHaveBeenCalledWith('/settings')
  })

  it('Back Button pushes to attest /profile page', async () => {
    mockKeycloak.hasRealmRole.mockReturnValue(false)
    const attestMenu = shallow(<AttestMenu />)
    const button = attestMenu.find('#attest-menu-back-button')
    button.simulate('click')
    expect(mockPush).toHaveBeenCalledTimes(1)
    expect(mockPush).toHaveBeenCalledWith('/profile')
  })
})
