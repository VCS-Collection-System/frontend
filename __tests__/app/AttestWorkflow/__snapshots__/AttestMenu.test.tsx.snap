// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttestWorkflow - AttestMenu should match existing snapshot 1`] = `
<Card
  isRounded={true}
>
  <CardWithTitle
    info="Please select from the following:"
    title="Menu"
  />
  <CardBody>
    <Stack
      hasGutter={true}
    >
      <Button
        icon={
          <FileUploadIcon
            color="currentColor"
            noVerticalAlign={false}
            size="sm"
          />
        }
        id="submit-vax-card-menu-button"
        isSmall={true}
        onClick={[Function]}
        variant="tertiary"
      >
        <Text>
          Submit Vaccination Record
        </Text>
        <Feature
          name="proofOptional"
          renderOff={[Function]}
        >
          Â 
        </Feature>
      </Button>
      <Feature
        name="covidTest"
        renderOff={[Function]}
      >
        <Button
          icon={
            <FileUploadIcon
              color="currentColor"
              noVerticalAlign={false}
              size="sm"
            />
          }
          id="submit-vax-test-menu-button"
          isSmall={true}
          onClick={[Function]}
          variant="tertiary"
        >
          <Text>
            Submit Test Results
          </Text>
          <Text>
            Supporting Document Upload Required
          </Text>
        </Button>
      </Feature>
    </Stack>
  </CardBody>
  <CardFooter>
    <Button
      id="attest-menu-back-button"
      onClick={[Function]}
      variant="plain"
    >
      Back
    </Button>
  </CardFooter>
</Card>
`;
